File: ta4j-core/src/main/java/org/ta4j/core/analysis/cost/LinearBorrowingCostModel.java
/*
 * SPDX-License-Identifier: MIT
 */
package org.ta4j.core.analysis.cost;

import java.util.Objects;
import org.ta4j.core.Position;
import org.ta4j.core.Trade;
import org.ta4j.core.Trade.TradeType;
import org.ta4j.core.num.Num;

/**
 * With this cost model, the trading costs for borrowing a position accrue
 * linearly.
 *
 * <p>
 * By default borrowing costs apply to short positions only, preserving the
 * historical behavior. Use
 * {@link #LinearBorrowingCostModel(double, Applicability)} to apply borrowing
 * costs to long positions, or to both long and short positions.
 */
public class LinearBorrowingCostModel implements CostModel {

    /**
     * Defines which entry sides incur borrowing costs.
     *
     * @since 0.23.1
     */
    public enum Applicability {

        /**
         * Borrowing costs apply to short positions only.
         *
         * @since 0.23.1
         */
        SHORT_ONLY,

        /**
         * Borrowing costs apply to long positions only.
         *
         * @since 0.23.1
         */
        LONG_ONLY,

        /**
         * Borrowing costs apply to long and short positions.
         *
         * @since 0.23.1
         */
        BOTH
    }

    /** The slope of the linear model (fee per period). */
    private final double feePerPeriod;

    /** Which position sides incur borrowing costs. */
    private final Applicability applicability;

    /**
     * Constructor with {@code feePerPeriod * nPeriod}. Borrowing costs apply to
     * short positions only.
     *
     * @param feePerPeriod the coefficient (e.g. 0.0001 for 1bp per period)
     */
    public LinearBorrowingCostModel(double feePerPeriod) {
        this(feePerPeriod, Applicability.SHORT_ONLY);
    }

    /**
     * Constructor with {@code feePerPeriod * nPeriod}.
     *
     * @param feePerPeriod  the coefficient (e.g. 0.0001 for 1bp per period)
     * @param applicability which position sides incur borrowing costs
     * @since 0.23.1
     */
    public LinearBorrowingCostModel(double feePerPeriod, Applicability applicability) {
        this.feePerPeriod = feePerPeriod;
        this.applicability = Objects.requireNonNull(applicability, "applicability must not be null");
    }

    /**
     * @return always {@code 0}, as borrowing costs depend on borrowed period
     */
    @Override
    public Num calculate(Num price, Num amount) {
        return price.getNumFactory().zero();
    }

    /**
     * @return the borrowing cost of the closed {@code position}
     * @throws IllegalArgumentException if {@code position} is still open
     */
    @Override
    public Num calculate(Position position) {
        if (position.isOpened()) {
            throw new IllegalArgumentException(
                    "Position is not closed. Final index of observation needs to be provided.");
        }
        return calculate(position, position.getExit().getIndex());
    }

    /**
     * @return the borrowing cost of the {@code position}
     */
    @Override
    public Num calculate(Position position, int currentIndex) {
        Trade entryTrade = position.getEntry();
        Trade exitTrade = position.getExit();
        Num borrowingCost = position.getEntry().getNetPrice().getNumFactory().zero();

        if (entryTrade != null && entryTrade.getAmount() != null && appliesTo(entryTrade.getType())) {
            int tradingPeriods = 0;
            if (position.isClosed()) {
                tradingPeriods = exitTrade.getIndex() - entryTrade.getIndex();
            } else if (position.isOpened()) {
                tradingPeriods = currentIndex - entryTrade.getIndex();
            }
            borrowingCost = getHoldingCostForPeriods(tradingPeriods, position.getEntry().getValue());
        }
        return borrowingCost;
    }

    private boolean appliesTo(TradeType tradeType) {
        return switch (applicability) {
        case BOTH -> true;
        case LONG_ONLY -> tradeType == TradeType.BUY;
        case SHORT_ONLY -> tradeType == TradeType.SELL;
        };
    }

    /**
     * @param tradingPeriods the number of periods
     * @param tradedValue    the value of the initial trading position of the trade
     * @return the absolute borrowing cost
     */
    private Num getHoldingCostForPeriods(int tradingPeriods, Num tradedValue) {
        return tradedValue.multipliedBy(tradedValue.getNumFactory()
                .numOf(tradingPeriods)
                .multipliedBy(tradedValue.getNumFactory().numOf(feePerPeriod)));
    }

    @Override
    public boolean equals(CostModel otherModel) {
        boolean equality = false;
        if (this.getClass().equals(otherModel.getClass())) {
            LinearBorrowingCostModel other = (LinearBorrowingCostModel) otherModel;
            equality = other.feePerPeriod == this.feePerPeriod && other.applicability == this.applicability;
        }
        return equality;
    }
}

File: ta4j-core/src/main/kotlin/org/ta4j/core/analysis/cost/LinearBorrowingCostModel.kt
/**
 * The MIT License (MIT)
 *
 * Copyright (c) 2017-2022 Ta4j Organization & respective
 * authors (see AUTHORS)
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy of
 * this software and associated documentation files (the "Software"), to deal in
 * the Software without restriction, including without limitation the rights to
 * use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of
 * the Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
 * FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
 * COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
 * IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
 * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */
package org.ta4j.core.analysis.cost

import org.ta4j.core.Position
import org.ta4j.core.Trade.TradeType
import org.ta4j.core.num.*

class LinearBorrowingCostModel
/**
 * Constructor. (feePerPeriod * nPeriod)
 *
 * @param feePerPeriod the coefficient (e.g. 0.0001 for 1bp per period)
 */(
    /**
     * Slope of the linear model - fee per period
     */
    private val feePerPeriod: Double
) : CostModel {
    override fun calculate(price: Num?, amount: Num?): Num {
        // borrowing costs depend on borrowed period
        return price!!.numOf(0)
    }

    /**
     * Calculates the borrowing cost of a closed position.
     *
     * @param position the position
     * @return the absolute trade cost
     */
    override fun calculate(position: Position): Num {
        require(!position.isOpened) { "Position is not closed. Final index of observation needs to be provided." }
        return calculate(position, position.exit!!.index)
    }

    /**
     * Calculates the borrowing cost of a position.
     *
     * @param position     the position
     * @param finalIndex final bar index to be considered (for open positions)
     * @return the absolute trade cost
     */
    override fun calculate(position: Position, finalIndex: Int): Num {
        val entryTrade = position.entry
        val exitTrade = position.exit
        var borrowingCost = position.entry!!.netPrice!!.numOf(0)

        // borrowing costs apply for short positions only
        if (entryTrade != null && entryTrade.type == TradeType.SELL) {
            var tradingPeriods = 0
            if (position.isClosed) {
                tradingPeriods = exitTrade!!.index - entryTrade.index
            } else if (position.isOpened) {
                tradingPeriods = finalIndex - entryTrade.index
            }
            borrowingCost = getHoldingCostForPeriods(tradingPeriods, position.entry!!.value!!)
        }
        return borrowingCost
    }

    /**
     * @param tradingPeriods number of periods
     * @param tradedValue    value of the trade initial trade position
     * @return the absolute borrowing cost
     */
    private fun getHoldingCostForPeriods(tradingPeriods: Int, tradedValue: Num): Num {
        return tradedValue
            .times(tradedValue.numOf(tradingPeriods).times(tradedValue.numOf(feePerPeriod)))
    }

    /**
     * Evaluate if two models are equal
     *
     * @param other model to compare with
     */
    override fun equals(other: Any?): Boolean {
        var equality = false
        if (this.javaClass == other!!.javaClass) {
            equality = (other as LinearBorrowingCostModel?)!!.feePerPeriod == feePerPeriod
        }
        return equality
    }

    companion object {
        private const val serialVersionUID = -2839623394737567618L
    }
}